src/gtc/common.py

```diff
@@ -816,15 +731,16 @@ def at_endpt(
             end=AxisBound(level=level, offset=end_offset),
         )
 
-    @root_validator
-    def check_start_before_end(cls, values: RootValidatorValuesType) -> RootValidatorValuesType:
-        if values["start"] and values["end"] and not (values["start"] <= values["end"]):
+    @datamodels.root_validator
+    @classmethod
```

Comment:
When I replace those two decorators with `@_make_root_validator` for example in `UnaryOp` I get the mypy error
```
src/gtc/common.py:415:5: error: The erased type of self "Type[gtc.common.UnaryOp[Any]]" is not a supertype of its class "gtc.common.UnaryOp[ExprT`1]"  [misc]
```
This is causing some post-traumatic stress disorder from the days of delicately working through the mypy issues here ;-). It's quite brittle. I had to already add more bound types inside eve (you can see by diffing the version in the branch compared to that in `functional`).

Reply:
So I'll keep this as-is for now, though we could open a ticket to see if there's a way to combine them. Probably low priority though, since we want to remove the `root_validator`s soon anyway.